Change generate_errors.pl to call perl grep

andresag01 opened the pull request:
Change the script generate_errors.pl to call the grep function in Perl
instead of calling the external tool grep directly as this causes
problems when ANSI escape sequences are included in the grep output
string.

This PR fixes https://github.com/ARMmbed/mbedtls/issues/504 and #1057

## Status
READY

## Requires Backporting
The problem is also present in mbedtls-2.1 (see https://github.com/armmbed/mbedtls/blob/mbedtls-2.1/scripts/generate_errors.pl#L49), but this is only a change in one of the tool scripts, so I do not think we should backport it. However, I am happy to do so if necessary.

## Todos
- [n/a] Tests
- [n/a] Documentation
- [n/a] Changelog updated
- [?] Backported (maybe)

## Steps to test or reproduce
Refer to original issue https://github.com/ARMmbed/mbedtls/issues/504

gilles-peskine-arm commented on scripts/generate_errors.pl:
```diff
@@ -46,7 +46,14 @@
 $/ = $line_separator;
 
-open(GREP, "grep \"define MBEDTLS_ERR_\" $include_dir/* |") || die("Failure when calling grep: $!");
+my @files = <$include_dir/*>;
```
While we're at it, this should really be `$include_dir/*.h`, to avoid catching editor backup files, patch rejections, etc.

gilles-peskine-arm replied:
While we're at it, shouldn't we at least try to exclude lines inside comments? Why doesn't the script even check if `define` is preceded by `#`?

sbutcher-arm commented on scripts/generate_errors.pl:
```diff
@@ -46,7 +46,14 @@
+my @matches;
+foreach my $file (@files) {
+    open(FILE, "$file");
+    my @grep_res = grep(/define MBEDTLS_ERR_/, <FILE>);
```
I agree with @gilles-peskine-arm. If we're going to change the script, we should also take the opportunity to make the pattern matching a little more robust.

andresag01 commented:
@gilles-peskine-arm @sbutcher-arm: I have modified the pattern matching a little so that it:
* Excludes all files that do not have the .h ending
* Excludes errors in single line comments and other not "well-formed" cases (such as using decimals instead of negative hex for the error)

However, note that the script cannot match errors defined within multi-line comments and I noticed there are some other places where the patterns are not as robust further down the script, but I don't think this is a problem as I changed the first match.

gilles-peskine-arm commented:
I've made a 2.1 backport because experience shows that when we don't backport changes to build/test scripts, we end up regretting it when it's time to backport the change we _do_ need, or to define CI jobs that work on every version.
